File: frontend/propmate-web/src/services/propertyService.js
const API_URL = "http://localhost:5235/api/properties";

// Get all properties with optional search/filter parameters
export async function getProperties(filters = {}) {
    const params = new URLSearchParams();

    if (filters.search) params.append("search", filters.search);
    if (filters.minPrice) params.append("minPrice", filters.minPrice);
    if (filters.maxPrice) params.append("maxPrice", filters.maxPrice);
    if (filters.bedrooms) params.append("bedrooms", filters.bedrooms);
    if (filters.bathrooms) params.append("bathrooms", filters.bathrooms);
    if (filters.location) params.append("location", filters.location);
    if (filters.isAvailable !== undefined && filters.isAvailable !== "") {
        params.append("isAvailable", filters.isAvailable);
    }

    const queryString = params.toString();

    const response = await fetch(
        queryString ? `${API_URL}?${queryString}` : API_URL
    );

    if (!response.ok) {
        throw new Error("Failed to load properties.");
    }

    return await response.json();
}

// Get one property by ID
export async function getPropertyById(id) {
    const response = await fetch(`${API_URL}/${id}`);

    if (!response.ok) {
        throw new Error("Property not found.");
    }

    return await response.json();
}

// Get property location
export async function getPropertyLocation(id) {
    const response = await fetch(`${API_URL}/${id}/location`);

    if (!response.ok) {
        throw new Error("Property location not found.");
    }

    return await response.json();
}
// Create a new property
export async function createProperty(property) {
    const response = await fetch(API_URL, {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify(property)
    });

    if (!response.ok) {
        throw new Error("Failed to create property.");
    }

    return await response.json();
}
// Update an existing property
export async function updateProperty(id, property) {
    const response = await fetch(`${API_URL}/${id}`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({
            id: Number(id),
            title: property.title,
            description: property.description,
            price: Number(property.price),
            bedrooms: Number(property.bedrooms),
            bathrooms: Number(property.bathrooms),
            location: property.location,
            latitude: Number(property.latitude),
            longitude: Number(property.longitude),
            isAvailable: property.isAvailable
        })
    });

    if (!response.ok) {
        throw new Error("Failed to update property.");
    }
}


// Delete a property
export async function deleteProperty(id) {
    const response = await fetch(`${API_URL}/${id}`, {
        method: "DELETE"
    });

    if (!response.ok) {
        throw new Error("Failed to delete property.");
    }
}

File: mobile/propmate_mobile/lib/services/property_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/property.dart';

class PropertyService {
  static const String baseUrl = 'http://localhost:5235/api/properties';

  Future<List<Property>> getProperties({
    String? search,
    double? minPrice,
    double? maxPrice,
    int? bedrooms,
    int? bathrooms,
    String? location,
    bool? isAvailable,
  }) async {
    final queryParameters = <String, String>{};

    if (search != null && search.trim().isNotEmpty) {
      queryParameters['search'] = search.trim();
    }

    if (minPrice != null) {
      queryParameters['minPrice'] = minPrice.toString();
    }

    if (maxPrice != null) {
      queryParameters['maxPrice'] = maxPrice.toString();
    }

    if (bedrooms != null) {
      queryParameters['bedrooms'] = bedrooms.toString();
    }

    if (bathrooms != null) {
      queryParameters['bathrooms'] = bathrooms.toString();
    }

    if (location != null && location.trim().isNotEmpty) {
      queryParameters['location'] = location.trim();
    }

    if (isAvailable != null) {
      queryParameters['isAvailable'] = isAvailable.toString();
    }

    final uri = Uri.parse(baseUrl).replace(
      queryParameters: queryParameters.isEmpty ? null : queryParameters,
    );

    final response = await http.get(uri);

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);

      return data.map((json) => Property.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load properties');
    }
  }

  Future<Property> getPropertyById(int id) async {
    final response = await http.get(Uri.parse('$baseUrl/$id'));

    if (response.statusCode == 200) {
      return Property.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load property');
    }
  }
}
